### PR TITLE
New resource: `azuredevops_workitemtrackingprocess_inherited_state`

### PR DESCRIPTION
**Pull request by Fresa**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [ ] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?
Add a new resource `azuredevops_workitemtrackingprocess_inherited_state`, that manages inherited work item type [state](https://learn.microsoft.com/en-us/rest/api/azure/devops/processes/states?view=azure-devops-rest-7.1) modifications.
image

image

Issue Number: #1488 

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Any relevant logs, error output, etc?
N/A

## Other information
Other unrelated unit tests are failing on main.

### CODE REVIEW

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+				Description:      "The ID of the process.",
+			},
+			"work_item_type_id": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotWhiteSpace),
+				Description:      "The ID (reference name) of the work item type.",
+			},
+			"name": {
```
Comment (Fresa) [inline azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:47]: We could use the ID here instead if we like. I chose name because it's simpler to identify than a UUID. State names are always unique within a WIT.

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotWhiteSpace),
+				Description:      "Name of the inherited state to manage.",
+			},
+			"visible": {
+				Type:        schema.TypeBool,
+				Optional:    true,
```
Comment (Fresa) [inline azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:56]: This could have been configured as required, but I chose to align it with all the other resources that have visibility configurable.

**Comment by Fresa** [comment]
@magodo this PR is ready for review, when you have the time!

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+		return nil, err
+	}
+
+	if states != nil {
+		for _, state := range *states {
+			if state.Name != nil && *state.Name == name {
+				if state.CustomizationType == nil {
+					return nil, fmt.Errorf("state %q has no customization type", name)
+				}
+				if *state.CustomizationType == workitemtrackingprocess.CustomizationTypeValues.Custom {
```
Comment (magodo) [inline azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:226]: Shouldn't we look for a state whose CustomizationType equals to Inherited?

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+func updateResourceInheritedState(ctx context.Context, d *schema.ResourceData, m any) diag.Diagnostics {
+	clients := m.(*client.AggregatedClient)
+
+	stateId := d.Id()
+	processId := d.Get("process_id").(string)
+	witRefName := d.Get("work_item_type_id").(string)
+
+	visible, err := getBoolAttributeFromConfig(d, "visible")
```
Comment (magodo) [inline azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:157]: Maybe we shall consider to modify this function `getBoolAttributeFromConfig` to a generic function that works for other types than boolean? Though it shall be done in another PR.

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go (hunk 1 above)
Comment (magodo) [reply azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:47]: Since the other related resources all use id, shall we also use id here to be consistent?

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+				WitRefName:     &witRefName,
+				StateId:        converter.UUID(stateId),
+				HideStateModel: &workitemtrackingprocess.HideStateModel{Hidden: &hidden},
+			})
+			if err != nil {
+				return diag.Errorf("hiding state: %+v", err)
+			}
+		} else {
+			// Use DELETE to unhide a state
+			err := clients.WorkItemTrackingProcessClient.DeleteStateDefinition(ctx, workitemtrackingprocess.DeleteStateDefinitionArgs{
```
Comment (magodo) [inline azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:176]: Assuming deleting it here won't really delete the state definition but just revert it, and it won't impact a follow up `GetStateDefinition` call?

File: azuredevops/internal/acceptancetests/resource_workitemtrackingprocess_inherited_state_test.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+			StateId:    &stateUUID,
+		})
+		if err != nil {
+			return fmt.Errorf("getting state definition: %w", err)
+		}
+
+		if state == nil {
+			return fmt.Errorf("inherited state should still exist after removing from Terraform but it was not found")
+		}
+
```
Comment (magodo) [inline azuredevops/internal/acceptancetests/resource_workitemtrackingprocess_inherited_state_test.go:193]: Shall we also test the visibility here is null once it was deleted?

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go (hunk 3 above)
Comment (Fresa) [reply azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:226]: "Inherited" states are actually labeled as "System" for some reason, when hiding a "System" state it turns into an "Inherited" state. The [Hide operation](https://learn.microsoft.com/en-us/rest/api/azure/devops/processes/states/hide-state-definition?view=azure-devops-rest-7.1&tabs=HTTP) only works on "System" states. I've never seen an "Inherited" state on an inherited work item type, except when hiding it. One could argue that we should call it `azuredevops_workitemtrackingprocess_system_state` instead, I'm not sure what makes most sense since it switches between them. System/inherited states only seem to exist on inherited work item types, and the description in the States tab in the UI (see screenshot in PR description above) refers to it as "inherited" state, probably because it is in fact inherited from the parent WIT.

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go (hunk 1 above)
Comment (Fresa) [reply azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:47]: It would simplify the implementation a bit as well. The IDs seems to be hardcoded globally, so predictable-ish, but would still require the user to find it via the state list endpoint or WIT GET endpoint. I don't have a strong opinion for either, you choose!

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go (hunk 5 above)
Comment (Fresa) [reply azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:176]: Yepp, exactly. 

If trying to use the Hide endpoint with `Hidden: false`, it will error and explicitly suggest using the DELETE endpoint for reverting.

File: azuredevops/internal/acceptancetests/resource_workitemtrackingprocess_inherited_state_test.go (hunk 6 above)
Comment (Fresa) [reply azuredevops/internal/acceptancetests/resource_workitemtrackingprocess_inherited_state_test.go:193]: Good point, done!

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go (hunk 3 above)
Comment (magodo) [reply azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:226]: Well, thanks for the context! Then could you add some comments here to clarify?

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go (hunk 1 above)
Comment (magodo) [reply azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:47]: ~With a further look, I would suggest we expand the `azuredevops_work_item_type` to expose the `states` (similar as the `pages`). So that the user can further convert the array of `states` to a map keyed by the state name, then reference the state id by the name. WDYT?~

For the UX, maybe we shall just keep the `name`, which the user can easily learn from the document/UI?

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go (hunk 3 above)
Comment (Fresa) [reply azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:226]: done!

File: azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go (hunk 1 above)
Comment (Fresa) [reply azuredevops/internal/service/workitemtrackingprocess/resource_inherited_state.go:47]: I think that makes sense

**Review by magodo — APPROVED** [review]
LGTM!

```shell
terraform-provider-azuredevops on  manage-inherited-states via 🐹 v1.25.6
💤 TF_ACC=1 go test -v -run='TestAccWorkitemtrackingprocessInheritedState_' ./azuredevops/internal/acceptancetests
=== RUN   TestAccWorkitemtrackingprocessInheritedState_Basic
=== PAUSE TestAccWorkitemtrackingprocessInheritedState_Basic
=== RUN   TestAccWorkitemtrackingprocessInheritedState_Update
=== PAUSE TestAccWorkitemtrackingprocessInheritedState_Update
=== RUN   TestAccWorkitemtrackingprocessInheritedState_RemoveFromState
=== PAUSE TestAccWorkitemtrackingprocessInheritedState_RemoveFromState
=== CONT  TestAccWorkitemtrackingprocessInheritedState_Basic
=== CONT  TestAccWorkitemtrackingprocessInheritedState_RemoveFromState
=== CONT  TestAccWorkitemtrackingprocessInheritedState_Update
--- PASS: TestAccWorkitemtrackingprocessInheritedState_Basic (3.72s)
--- PASS: TestAccWorkitemtrackingprocessInheritedState_Update (3.94s)
--- PASS: TestAccWorkitemtrackingprocessInheritedState_RemoveFromState (5.62s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        5.624s
```
